core/src/main/java/org/infinispan/scattered/impl/ScatteredVersionManagerImpl.java

```diff
@@ -267,15 +270,21 @@ public boolean isVersionActual(int segment, EntryVersion version) {
    }
 
    @Override
-   public void notifyKeyTransferFinished(int segment, boolean expectValues) {
-      SegmentState update = expectValues ? SegmentState.VALUE_TRANSFER : SegmentState.OWNED;
-      SegmentState previous;
-      do {
-         previous = segmentStates.get(segment);
-         // It is possible that the segment is not in KEY_TRANSFER state, but can be in BLOCKED states as well
-         // when the CONFIRM_REVOKED_SEGMENTS failed.
-         log.tracef("Finishing transfer for segment %d = %s -> %s", segment, previous, update);
-      } while (!segmentStates.compareAndSet(segment, previous, update));
+   public void notifyKeyTransferFinished(int segment, boolean expectValues, boolean cancelled) {
+      SegmentState update;
+      if (expectValues) {
+         update = SegmentState.VALUE_TRANSFER;
+      } else if (cancelled) {
```

Comment:
`expectValues` is `true` only if the transfer was not cancelled, see `ScatteredStateConsumerImpl:231`. Though, I admit that this is a risk for the future.